src/goto-instrument/havoc_utils.cpp

```diff
@@ -92,3 +84,29 @@ void append_havoc_code(
     append_scalar_havoc_code_for_expr(source_location, expr, dest);
   }
 }
+
+void append_havoc_code_if_valid(
+  const source_locationt source_location,
+  const namespacet &ns,
+  const modifiest &modifies,
+  goto_programt &dest)
+{
+  havoc_utils_is_constantt is_constant(modifies);
+  for(const auto &expr : modifies)
+  {
+    if(expr.id() == ID_index || expr.id() == ID_dereference)
+    {
+      address_of_exprt address_of_expr(expr);
+      if(!is_constant(address_of_expr))
+      {
+        append_safe_havoc_code_for_expr(source_location, ns, expr, dest, [&]() {
+          append_object_havoc_code_for_expr(source_location, expr, dest);
+        });
+        continue;
```

Comment:
Lack of test coverage on this conditional is a concern

Reply:
Good point. We do have a test case for multi-dimensional decreases clauses, and it seems to be working as expected. I can take a closer look at it, but I will probably split any fixes (if needed) into a separate PR, and keep this one only about validity checking.